File: scripts/SDL_Interface/PythonGenerator/PythonMethod.py

class PythonMethod:

	UNIT_INDENT = "    "

	def __init__(self, name=""):
		self.name = name
		self.inputs = ["self"]
		self.content_lines = []

	def set_name(self, name):
		self.name = name

	# TODO: use keyword arg, otherwise we cannot default to None
	def add_input(self, name, **kwargs):
		input_string = name
		if "expected_type" in kwargs:
			input_string += ": " + kwargs["expected_type"]
		elif "default_value" in kwargs:
			input_string += " = " + kwargs["default_value"]
		self.inputs.append(input_string)

	def add_content_line(self, content):
		self.content_lines.append(PythonMethod.UNIT_INDENT + content + "\n")

	def get_name(self):
		return self.name

	def gen_code(self, indent_amount=0):

		code = ""

		if not self.name:
			print("warning: python method has no name")
			return code

		input_list = ", ".join(self.inputs)
		indention = PythonMethod.UNIT_INDENT * indent_amount

		code += "%sdef %s(%s):\n" % (indention, self.name, input_list)

		if self.content_lines:
			for line in self.content_lines:
				code += "%s%s" % (indention, line)
		else:
			code += indention + PythonMethod.UNIT_INDENT + "pass\n"

		return code + "\n"

File: SDLGenerator/SDLGen/Source/InterfaceGen/Python/PythonMethod.cpp
#include "InterfaceGen/Python/PythonMethod.h"

#include <Common/assertion.h>
#include <Common/logging.h>
#include <Common/Utility/string_utils.h>

#include <utility>

namespace ph::sdlgen
{

PH_DEFINE_INTERNAL_LOG_GROUP(PythonMethod, SDLGen);

PythonMethod::PythonMethod(std::string methodName)
	: m_methodName(std::move(methodName))
	, m_codeLines()
	, m_inputs()
	, m_currentIndentAmount(0)
{
	PH_ASSERT(!m_methodName.empty());

	addInput("self");
}

void PythonMethod::setDoc(std::string doc)
{
	m_doc = doc;
}

void PythonMethod::addInput(
	std::string variableName,
	std::string defaultArgument,
	std::string expectedType)
{
	std::string inputString = std::move(variableName);

	if(!expectedType.empty())
	{
		inputString += ": " + expectedType;
	}

	if(!defaultArgument.empty())
	{
		inputString += " = " + defaultArgument;
	}

	m_inputs.push_back(std::move(inputString));
}

void PythonMethod::addCodeLine(std::string codeLine)
{
	// Code line in a method should be indented too (hence indent amount +1)
	const std::string indention = string_utils::repeat(UNIT_INDENT, m_currentIndentAmount + 1);

	m_codeLines.push_back(
		std::format("{}{}\n", indention, std::move(codeLine)));
}

void PythonMethod::beginIndent()
{
	++m_currentIndentAmount;
}

void PythonMethod::endIndent()
{
	PH_ASSERT_GT(m_currentIndentAmount, 0);

	--m_currentIndentAmount;
}

std::string PythonMethod::genCode(const std::size_t indentAmount) const
{
	std::string code;

	if(m_methodName.empty())
	{
		PH_LOG(PythonMethod, Warning, "no method name provided, ignoring");
		return "";
	}

	std::string inputList;
	if(!m_inputs.empty())
	{
		inputList += m_inputs[0];
	}
	for(std::size_t i = 1; i < m_inputs.size(); ++i)
	{
		inputList += ", " + m_inputs[i];
	}

	const std::string indention = string_utils::repeat(UNIT_INDENT, indentAmount);

	code += std::format("{}def {}({}):\n", 
		indention, m_methodName, inputList);

	if(!m_doc.empty())
	{
		code += std::format("{}{}\"\"\"\n{}{}{}\n{}{}\"\"\"\n",
			indention, UNIT_INDENT, indention, UNIT_INDENT, m_doc, indention, UNIT_INDENT);
	}

	if(!m_codeLines.empty())
	{
		for(const std::string& codeLine : m_codeLines)
		{
			code += std::format("{}{}",
				indention, codeLine);
		}
	}
	else
	{
		code += std::format("{}{}pass\n",
			indention, UNIT_INDENT);
	}

	return code + "\n";
}

}// end namespace ph::sdlgen
